lib/pharos/root_command.rb

```diff
@@ -8,7 +8,7 @@
 require_relative 'terraform_command'
 
 module Pharos
-  class RootCommand < Pharos::Command
+  class RootCommand < Clamp::Command
```

Comment:
Subcommand wrapper classes cannot inherit from `Pharos::Command` because it will mess up exception handling.

Reply:
That breaks `pharos --version. How does it break exception handling?

Reply:
`super` gets called twice, other without options and it breaks the exception handling.

Reply:
Hmm, I don't see how that would happen

```console
# up command modified to raise on execute and commands to report super:
$ bundle exec bin/pharos up -c buster.yml
run of RootCommand, [["up", "-c", "buster.yml"]]
run of Pharos::Command [["up", "-c", "buster.yml"]]
parse of RootCommand [["up", "-c", "buster.yml"]]
parse of Pharos::Command [["up", "-c", "buster.yml"]]
run in UpCommand [["-c", "buster.yml"]]
run of Pharos::Command [["-c", "buster.yml"]]
parse of UpCommand [["-c", "buster.yml"]]
parse of Pharos::Command [["-c", "buster.yml"]]
rescue 4 of Pharos::Command RuntimeError in Pharos::UpCommand # signal_error raises Clamp::ExecutionError
rescue 1 of Pharos::Command Clamp::ExecutionError in Pharos::RootCommand 
ERROR: RuntimeError : foooooo!
```

```diff
diff --git a/lib/pharos/command.rb b/lib/pharos/command.rb
index ffeeda8..e50314f 100644
--- a/lib/pharos/command.rb
+++ b/lib/pharos/command.rb
@@ -12,20 +12,29 @@ module Pharos
       end
     end
 
-    def run(*_args)
-      Pharos::CoreExt::Colorize.disable! unless color?
+    def parse(*opts)
+      puts "parse of Pharos::Command #{opts.inspect}"
       super
-    rescue Clamp::HelpWanted, Clamp::ExecutionError, Clamp::UsageError
+    end
+
+    def run(*args)
+      puts "run of Pharos::Command #{args.inspect}"
+      super
+    rescue Clamp::HelpWanted, Clamp::ExecutionError, Clamp::UsageError => ex
+      puts "rescue 1 of Pharos::Command #{ex.class.name} in #{self.class.name}"
       raise
     rescue Errno::EPIPE => ex
+      puts "rescue 2 of Pharos::Command #{ex.class.name} in #{self.class.name}"
       raise if debug?
 
       warn "ERROR: #{ex.class.name} : #{ex.message}" if $stdout.tty?
       exit 141
-    rescue Pharos::ConfigError => exc
-      warn "==> #{exc}"
+    rescue Pharos::ConfigError => ex
+      puts "rescue 3 of Pharos::Command #{ex.class.name} in #{self.class.name}"
+      warn "==> #{ex}"
       exit 11
     rescue StandardError => ex
+      puts "rescue 4 of Pharos::Command #{ex.class.name} in #{self.class.name}"
       raise if debug?
 
       signal_error "#{ex.class.name} : #{ex.message}"
@@ -42,6 +51,7 @@ module Pharos
       ENV["DEBUG"] = "true"
     end
 
+
     def prompt
       @prompt ||= TTY::Prompt.new(enable_color: color?)
     end
diff --git a/lib/pharos/root_command.rb b/lib/pharos/root_command.rb
index 6816786..352125c 100644
--- a/lib/pharos/root_command.rb
+++ b/lib/pharos/root_command.rb
@@ -17,5 +17,19 @@ module Pharos
     subcommand %w(exec ssh), "run a command or an interactive session on a host", ExecCommand
     subcommand "tf", "terraform specific commands", TerraformCommand
     subcommand "version", "show version information", VersionCommand
+
+    def parse(*opts)
+      puts "parse of RootCommand #{opts.inspect}"
+      super
+    end
+
+    def run(*args)
+      puts "run of RootCommand, #{args.inspect}"
+      super
+    end
+
+    def execute
+      puts "in execute of RootCommand"
+    end
   end
 end
diff --git a/lib/pharos/up_command.rb b/lib/pharos/up_command.rb
index bab48a8..5cb1fbb 100644
--- a/lib/pharos/up_command.rb
+++ b/lib/pharos/up_command.rb
@@ -8,7 +8,18 @@ module Pharos
 
     option ['-f', '--force'], :flag, "force upgrade"
 
+    def parse(*opts)
+      puts "parse of UpCommand #{opts.inspect}"
+      super
+    end
+
+    def run(*args)
+      puts "run in UpCommand #{args.inspect}"
+      super
+    end
+
     def execute
+      raise "foooooo!"
       puts "==> KONTENA PHAROS v#{Pharos.version} (Kubernetes v#{Pharos::KUBE_VERSION})".bright_green
 
       Pharos::Kube.init_logging!
```

Reply:
If I add "debug" to the rescue it will output this:
```
"class => Pharos::UpCommand"
"debug => true"
"class => Pharos::RootCommand"
"debug => nil"
ERROR: RuntimeError : foo
```

Reply:
Ah, that's not `super`

1) RootCommand sees `up` keyword and spawns an `UpCommand` instance with the remaining arguments
2) RootCommand extends `Clamp::Subcommand::Execution` on itself, which replaces `execute`:
```ruby
      def execute
        # delegate to subcommand
        subcommand = instantiate_subcommand(subcommand_name)
        subcommand.run(subcommand_arguments)
      end
```
2) `UpCommand#parse` consumes `--debug`
3) `subcommand.run` raises inside RootCommand.
4) `RootCommand#run`'s `rescue` is triggered, RootCommand doesn't know / care about `--debug` because `UpCommand` consumed it, so it's `nil`.

1: UpCommand does:
```ruby
rescue StandardError => ex
  raise if debug? # it's true here, so it is reraised
  signal_error "#{ex.class.name} : #{ex.message}"
end
```

2: Which goes into `RootCommand < Pharos::Command` rescue:
```ruby
rescue StandardError => ex
  raise if debug? # it's nil, so it does not raise here
  signal_error "#{ex.class.name} : #{ex.message}" # raises Clamp::ExecutionError
end
```

If you use `Clamp::Command` instead of `Pharos::Command`:

1: UpCommand does:
```ruby
rescue StandardError
  raise if debug?  
  signal_error "whoops"
end
```

2: Which goes into `RootCommand < Clamp::Command`:
```ruby
rescue Clamp::UsageError => e
  $stderr.puts "ERROR: #{e.message}"
  $stderr.puts ""
  $stderr.puts "See: '#{e.command.invocation_path} --help'"
  exit(1)
rescue Clamp::HelpWanted => e
  puts e.command.help
rescue Clamp::ExecutionError => e
  $stderr.puts "ERROR: #{e.message}"
  exit(e.status)
rescue SignalException => e
  exit(128 + e.signo)
end
```

So replacing it with `Clamp::Command` does make a difference. In the other PR it's handled by setting `Pharos::Logging.debug = true` and using the global `Pharos::Logging.debug?` instead of `debug?`